### PR TITLE
SNOW-968741: Arrow performance optimizations

### PR DESCRIPTION
**Pull request by sfc-gh-pbulawa**
### Description
Optimizations for arrow processing.

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [x] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data/Core/ArrowResultChunk.cs [hunk 1]
```diff
@@ -123,106 +164,135 @@ public object ExtractCell(int columnIndex, SFDataType srcType, long scale)
...
+                        _int[columnIndex] = ((Date32Array)column).Values.ToArray();
+                    return SFDataConverter.UnixEpoch.AddTicks(_int[columnIndex][_currentRecordIndex] * TicksPerDay);
+                
                 case SFDataType.TIME:
                 {
-                    var value = column.Data.DataType.TypeId == ArrowTypeId.Int32
-                        ? ((Int32Array)column).GetValue(_currentRecordIndex)
-                        : ((Int64Array)column).GetValue(_currentRecordIndex);
+                    var value = column.GetType() == typeof(Int32Array)
+                        ? ((Int32Array)column).Values[_currentRecordIndex]
```
Comment (sfc-gh-dstempniak) [inline Snowflake.Data/Core/ArrowResultChunk.cs:241]: I think it should be also replaced by cached values because Values property is not an array, it is implemented as
`public ReadOnlySpan<T> Values => ValueBuffer.Span.CastTo<T>().Slice(Offset, Length);`
so it is not efficient.

File: Snowflake.Data/Core/ArrowResultChunk.cs [hunk 2]
```diff
@@ -111,6 +144,14 @@ public override UTF8Buffer ExtractCell(int columnIndex)
             throw new NotSupportedException();
         }
 
+        public string GetMetadata()
```
Comment (sfc-gh-dstempniak) [inline Snowflake.Data/Core/ArrowResultChunk.cs:147]: I would remove this function, it was for testing purpose.

Comment (sfc-gh-pbulawa) [reply Snowflake.Data/Core/ArrowResultChunk.cs]: Done

File: Snowflake.Data/Core/ArrowResultChunk.cs (hunk 1 above)
Comment (sfc-gh-pbulawa) [reply Snowflake.Data/Core/ArrowResultChunk.cs]: Done

**Review by sfc-gh-dstempniak — APPROVED** [review]
LGTM

File: Snowflake.Data/Core/ArrowResultSet.cs [hunk 3]
```diff
@@ -60,8 +44,28 @@ public ArrowResultSet(QueryExecResponseData responseData, SFStatement sfStatemen
                 isClosed = false;
 
                 queryId = responseData.queryId;
+                
+                if (responseData.rowsetBase64.Length > 0)
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data/Core/ArrowResultSet.cs:48]: let's extract it into a metod.For example something like: _currentChunk = ReadChunk(...)

File: Snowflake.Data/Core/ArrowResultChunk.cs [hunk 4]
```diff
@@ -123,106 +156,174 @@ public object ExtractCell(int columnIndex, SFDataType srcType, long scale)
...
+                        case Int8Array array:
+                            if (_sbyte[columnIndex] == null)
+                                _sbyte[columnIndex] = array.Values.ToArray();
                             if (scale == 0)
-                                return ((Int8Array)column).GetValue(_currentRecordIndex);
-                            else
-                                return ((Int8Array)column).GetValue(_currentRecordIndex) / (decimal)s_powersOf10[scale];
-                        case ArrowTypeId.Int16:
+                                return _sbyte[columnIndex][_currentRecordIndex];
+                            return _sbyte[columnIndex][_currentRecordIndex] / (decimal)s_powersOf10[scale];
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data/Core/ArrowResultChunk.cs:166]: maybe some tests for the null values could be added

File: Snowflake.Data/Core/BaseResultChunk.cs [hunk 5]
```diff
@@ -32,12 +36,16 @@ public abstract class BaseResultChunk : IResultChunk
...
         
         internal abstract bool Rewind();
-
+        
         internal virtual void Reset(ExecResponseChunk chunkInfo, int chunkIndex)
         {
             RowCount = chunkInfo.rowCount;
             Url = chunkInfo.url;
             ChunkIndex = chunkIndex;
+
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data/Core/BaseResultChunk.cs:45]: let's remove empty lines

File: Snowflake.Data/Core/BaseResultChunk.cs [hunk 6]
```diff
@@ -32,12 +36,16 @@ public abstract class BaseResultChunk : IResultChunk
...
+        
         internal virtual void Reset(ExecResponseChunk chunkInfo, int chunkIndex)
         {
             RowCount = chunkInfo.rowCount;
             Url = chunkInfo.url;
             ChunkIndex = chunkIndex;
+            
+            CompressedSize = chunkInfo.compressedSize;
+            UncompressedSize = chunkInfo.uncompressedSize;
+
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data/Core/BaseResultChunk.cs:48]: and here

File: Snowflake.Data/Core/BaseResultChunk.cs (hunk 5 above)
Comment (sfc-gh-pbulawa) [reply Snowflake.Data/Core/BaseResultChunk.cs:45]: Done

File: Snowflake.Data/Core/BaseResultChunk.cs (hunk 6 above)
Comment (sfc-gh-pbulawa) [reply Snowflake.Data/Core/BaseResultChunk.cs:48]: Done

File: Snowflake.Data/Core/ArrowResultChunk.cs (hunk 4 above)
Comment (sfc-gh-pbulawa) [reply Snowflake.Data/Core/ArrowResultChunk.cs:166]: Added more tests for null records.

File: Snowflake.Data/Core/ArrowResultSet.cs (hunk 3 above)
Comment (sfc-gh-pbulawa) [reply Snowflake.Data/Core/ArrowResultSet.cs]: Done

**Review by sfc-gh-knozderko — APPROVED** [review]
